File: backend/app/schemas/task.py
"""Pydantic schemas for Task and TaskStep — shared contracts for the Task System."""

from datetime import datetime
from enum import StrEnum
from typing import Optional

from pydantic import BaseModel, ConfigDict


class TaskStatus(StrEnum):
    pending = "pending"
    in_progress = "in_progress"
    done = "done"


class TaskStepCreate(BaseModel):
    title: str
    order: int


class TaskStepRead(TaskStepCreate):
    id: int
    task_id: int
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.pending


class TaskRead(TaskCreate):
    id: int
    user_id: int
    steps: list[TaskStepRead] = []
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[TaskStatus] = None

File: frontend/src/types/task.ts
/**
 * Frontend type definitions for Task system.
 * Must match backend schemas/task.py (TaskRead, TaskCreateRequest, TaskUpdateRequest).
 */

export enum TaskStatus {
  PENDING = 'PENDING',
  PROCESSING = 'PROCESSING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  CANCELLED = 'CANCELLED',
  RETRYING = 'RETRYING',
}

export interface TaskStep {
  id: string;
  task_id: string;
  step_index: number;
  step_type: string;
  agent_name: string;
  status: string;
  model_used: string | null;
  tokens_in: number | null;
  tokens_out: number | null;
  latency_ms: number | null;
  confidence: number | null;
  output_payload: Record<string, unknown> | null;
  created_at: string;
  completed_at: string | null;
}

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  task_type: string | null;
  priority: number;
  retry_count: number;
  config: Record<string, unknown> | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  result: Record<string, unknown> | null;
  error: Record<string, unknown> | null;
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: number;
  config?: Record<string, unknown>;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: number;
  config?: Record<string, unknown>;
}

export interface TaskStatusResponse {
  id: string;
  status: TaskStatus;
}

export interface TaskDetailResponse extends Task {
  steps: TaskStep[];
}
